lib/portlayer/exec2/remote/server/portlayer_rpc_server.go

```diff
@@ -0,0 +1,143 @@
+// Copyright 2016 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"encoding/hex"
+	"fmt"
+	"log"
+	"math/rand"
+	"net"
+	"net/http"
+	"net/rpc"
+
+	"github.com/vmware/vic/lib/portlayer/exec2"
+	"github.com/vmware/vic/lib/portlayer/exec2/remote"
+)
+
+type PortLayerRPCServer struct {
+}
+
+// A Handle can be anything, so this takes advantage of that by creating a sparse handle
+// to send to the client and using that sparse handle as a key in a hashtable which points to
+// rich handles created by the HandleFactory
+type SparseHandle exec2.Handle
+
+var lcTarget exec2.ContainerLifecycle
+
+//var lcQuery exec2.ContainerQuery
+var handles map[SparseHandle]exec2.Handle
```

Comment:
Good question Hasan. The code already kind of is the _stateless_ design you're describing. The HandleFactory abstraction allows us to control where the handle is created and the handle is never stored anywhere in the PL. In this respect, portlayer_rpc_server.go is a client. It controls the HandleFactory passed in and it effectively "owns" the structs that are passed in and out of the PL. In this incarnation, the handle is a pointer to a struct created by the client. However, in the case where we're remoting, passing rich data structs around might be kind of clumsy. We can choose to do that if we really want, but the handle means that we don't have to. This is what I mean by the distinction between rich and sparse handles. The interface is opaque enough that the client doesn't need to know the difference as it will never try to modify what the handle points to directly. However, the question of what the handle points to is important because it's the scope of that data that matters.